[Coop thread threadpool mmap] Convert more functions to coop/handle.

jaykrell opened the pull request:
[Coop] Convert mono_mmap_open_file, mono_mmap_open_handle to coop/handle.
[Coop] Convert ves_icall_System_Threading_Thread_Yield to coop/handle (no pointers).
[Coop] Convert ves_icall_System_Threading_Thread_current_lcid to coop/handle -- no pointers.
[Coop] Rename some error to ioerror or win32error.
[Coop] Convert mmap to coop/handle -- where there aren't pointers, just unused errors.
[Coop] Convert ves_icall_System_Threading_InternalThread_Thread_free_internal to coop/handle.
[Coop] Convert threadpool.c to handles -- no pointers, just MonoError.
[Coop] Convert ves_icall_System_Threading_Thread_SpinWait_nop to coop/handle.
[Coop] Convert ves_icall_System_Threading_Thread_Sleep_internal to coop/handle.
[Coop] Convert ves_icall_System_Threading_Thread_Suspend to coop/handle.
[Coop] Add mono_set_pending_exception_handle and fix indent.
[Coop] Convert mono_threads_get_max_stack_size ves_icall_System_Threading_Thread_SystemMaxStackSize/typo SystemMaxStackStize to coop/handle.
[Coop] Convert ves_icall_System_Threading_Thread_Thread_internal to coop/handle.
[Coop] Convert ves_icall_System_Threading_Thread_ConstructInternalThread to coop/handle.

https://github.com/mono/mono/issues/6851

jaykrell commented:
@monojenkins build Linux x64 Coop GC

jaykrell commented:
review?

jaykrell commented:
build

jaykrell commented:
@monojenkins build Linux i386 Coop GC

Therzok commented on mono/metadata/file-mmap-windows.c:
```diff
@@ -132,66 +132,71 @@ static int convert_win32_error (int error, int def)
 }
 
-static void *open_handle (void *handle, MonoString *mapName, int mode, gint64 *capacity, int access, int options, int *error)
+<<<<<<< HEAD
```
Conflict markers.

luhenry commented on mono/metadata/file-mmap.h:
```diff
@@ -18,18 +18,25 @@
 
-extern void *mono_mmap_open_file (MonoString *string, int mode, MonoString *mapName, gint64 *capacity, int access, int options, int *error);
+void*
+mono_mmap_open_file (MonoStringHandle string, int mode, MonoStringHandle mapName, gint64 *capacity, int access, int options, int *win32error, MonoError *error);
```
It would be simpler to use `fixed` in managed.

jaykrell commented:
I was erring toward smaller change but agreed more fixed, less C/C++ is good.
Why don't we just, like, fixed everything? I mean, just make the raw pointers to all types just work, but putting in fixed at the managed layer?  For strings, fixed would yield a char* but we'd still just pass the object. Of course this doesn't quite handle return values or exceptions, but I kinda think you can greatly accelerate coop conversion by sprinking fixed in all the callers. For out parameters -- is there some sort of "fixed reference" or pass the address of a local? And do that for return values?

luhenry commented on mcs/class/System.Core/System.IO.MemoryMappedFiles/MemoryMappedFile.cs:
```diff
@@ -93,19 +93,38 @@ internal static void Map (IntPtr handle, long offset, ref long size, MemoryMappe
+			// Underlying implementation does not handle embedded nuls.
+			if (s == null || s.IndexOf((char)0) != -1)
+				return;
+			throw new ArgumentException ($"String \"{s}\" contains embedded nul.");
```
This changes the behavior of the API. Why is it needed?

jaykrell replied:
I agree it is separate, maybe a separate PR? The preexisting behavior was faulty -- if presented with string with embedded nuls it would silently truncate. Trying to open foo\0bar should fail and not silently open foo.

jaykrell commented:
review again?

jaykrell commented:
ping?

jaykrell commented:
Oh it is backwards looking, isn't? I'll undo it here make a separate PR for it.

jaykrell commented:
ok, moved that change to a separate PR. ping?
This does overlap/conflict somewhat with another PR but I'm ok with resolving that. They are both really the same, esp. around Thread.Sleep, but the other PR is more complete.

jaykrell commented:
@monojenkins build  Linux x64 mcs

jaykrell commented:
Ping?

jaykrell commented:
Ping?

lambdageek commented on mono/metadata/file-mmap.h:
```diff
@@ -18,18 +18,25 @@
 #include <mono/utils/mono-compiler.h>
 
-extern void mono_mmap_close (void *mmap_handle);
+void
```
I don't know if it's alright to remove `extern` from these icall declarations.

jaykrell replied:
Extern is pretty meaningless, noise, just messes up the code.

lambdageek commented on mono/metadata/threads.c:
```diff
@@ -1434,19 +1445,30 @@ ves_icall_System_Threading_Thread_GetCurrentThread (MonoError *error)
+	internal = thread_handle_to_internal_ptr (thread_handle);
+
+	if (!internal) {
+		ves_icall_System_Threading_Thread_ConstructInternalThread (thread_handle, error);
```
Just personal preference, but I like to keep icall entrypoints separate from internal utility functions that are called from elsewhere in the runtime so that if we ever change what happens at the icall boundary we don't have to constrain callers elsewhere in the runtime in the same way.

jaykrell replied:
Sure, but that is how it was.

jaykrell replied:
Fixing this..

lambdageek commented on mono/metadata/threads.c:
```diff
@@ -1434,19 +1445,30 @@ ves_icall_System_Threading_Thread_GetCurrentThread (MonoError *error)
+
+	if (!internal) {
+		ves_icall_System_Threading_Thread_ConstructInternalThread (thread_handle, error);
+		mono_error_assert_ok (error);
```
Why is this an assert ok?

jaykrell replied:
It appears before and after to assume success. I'd rather fail an assert than null dereference, perhaps -- there are arguments either way. The processor is always doing the checks for you, they do work, asserting redundantly all over the place is bloating. I can remove it.

jaykrell replied:
Removing "ves_icall" and the assert go together, since the underlying piece upon refactor never sets error, ok.

lambdageek commented on mono/metadata/threads.c:
```diff
@@ -1434,19 +1445,30 @@ ves_icall_System_Threading_Thread_GetCurrentThread (MonoError *error)
 MonoBoolean
-ves_icall_System_Threading_Thread_Thread_internal (MonoThread *this_obj,
-												   MonoObject *start)
+ves_icall_System_Threading_Thread_Thread_internal (MonoThreadObjectHandle thread_handle, MonoObjectHandle start_handle, MonoError *error)
```
lower in teh body of this icall we use `mono_set_pending_exception (mono_get_exception_thread_state (...))` - we should convert it to `mono_error_set_generic (error, "System.Threading", "ThreadStateException", msg)` (or even better - a new `mono_error_set_thread_state` that calls `mono_error_set_generic`).

jaykrell replied:
Let me look into that. I was following the pattern, but perhaps not a good pattern.

lambdageek commented on mono/metadata/threads.c:
```diff
@@ -1534,9 +1557,12 @@ ves_icall_System_Threading_Thread_Sleep_internal(gint32 ms)
-				mono_set_pending_exception (exc);
+			MonoExceptionHandle exc = mono_thread_execute_interruption_handle (error);
+			mono_error_assert_ok (error);
+			if (exc && !MONO_HANDLE_IS_NULL (exc)) {
```
a `MonoExceptionHandle` shouldn't be null.

jaykrell replied:
Low memory?

luhenry replied:
We don't care about low memory. It's never been a design choice we have made and so it will take much more than this check to make it work.

lambdageek commented on mono/metadata/threads.c:
```diff
@@ -1534,9 +1557,12 @@ ves_icall_System_Threading_Thread_Sleep_internal(gint32 ms)
+			MonoExceptionHandle exc = mono_thread_execute_interruption_handle (error);
+			mono_error_assert_ok (error);
+			if (exc && !MONO_HANDLE_IS_NULL (exc)) {
+				mono_set_pending_exception_handle (exc);
```
Instead of adding a variant of `mono_set_pending_exception` I think a better approach is to make a version of `mono_thread_execute_interruption` that sets a `MonoError` if there was an interruption.  We want fewer uses of `mono_set_pending_exception` in the runtime, not more.

jaykrell replied:
This code has proven very sensitive and fails upon making correct looking changes.

jaykrell replied:
I don't think this made things worse in this regard either but I can look again.

jaykrell replied:
The exception is not always set to pending so that will require a little more refactoring.

jaykrell replied:
This somewhat confuses an exception for interruption vs. an exception creating said exception.
The first is expected sometimes, the second is low memory and not expected -- but we do thread the error parameter every where and this is the main reason it can be set, even though we ignore low memory sometimes.

jaykrell replied:
This will maybe mess with lifetime. Which maybe is the problem I had changing this area in the first place. i.e. other than returning the exception, we also set thread->abort_exc and/or make it pending.

jaykrell replied:
I suggest we leave this asis for this PR. The code is a bit messy and I made a pretty minimal mechanical change.

lambdageek commented on mono/metadata/threads.c:
```diff
@@ -2554,12 +2579,10 @@ mono_thread_suspend (MonoInternalThread *thread)
-		return;
-	}
+	if (!mono_thread_suspend (thread_handle_to_internal_ptr (this_obj)))
+		mono_set_pending_exception_handle (mono_exception_new_thread_state ("Thread has not been started, or is dead.", error));
```
Same as above - just set the MonoError via `mono_error_set_generic`

jaykrell replied:
I was trying to stay close to original and didn't notice that option. I'll look.
extern is pretty much meaningless noise in C and C++ and should be generally omitted as such. Like "auto" (historically, not lately).

lambdageek commented on mono/metadata/threads.c:
```diff
@@ -4584,6 +4607,69 @@ mono_thread_execute_interruption (void)
+			goto exit;
+		exc = mono_exception_new_thread_abort (error);
+		if (!MONO_HANDLE_IS_NULL (exc)) {
+			thread->abort_exc = MONO_HANDLE_RAW (exc);
```
Isn't this missing a gc write barrier?  (the corresponding code in the version not using handles does a `MONO_OBJECT_SETREF` here)

jaykrell replied:
Maybe. It is within a lock, but all reads/writes would have to be for that to suffice and you are right it isn't "maintained in kind".

lambdageek commented on mono/metadata/threads.c:
```diff
@@ -1534,9 +1562,12 @@ ves_icall_System_Threading_Thread_Sleep_internal(gint32 ms)
-			MonoException* exc = mono_thread_execute_interruption ();
-			if (exc) {
-				mono_set_pending_exception (exc);
+			MonoExceptionHandle exc = mono_thread_execute_interruption_handle (error);
```
I'm still quite weary of adding new tech debt (`mono_thread_execute_interruption_handle` as a copy of all the logic of `mono_thread_execute_interruption`) for the sake of `Thread.Sleep_internal`.  Are we going to update all the other callers of `mono_thread_execute_interruption` to use the handle version in relatively short order?  If not, maybe we could either leave Sleep_internal out of this PR, or else change `mono_thread_execute_interruption_handle` to just be a wrapper around `mono_thread_execute_interruption` for now?
Having to maintain two copies of thread interruption in sync doesn't sound fun.

Otherwise, this LGTM.

jaykrell replied:
Yes.
I went back and forth on this.
I tried just layering pointer over handle.
That didn't work, for not understood reasons.

So I decided cloning the logic is easier/safer in the short term.

But ues it is also gross duplication.

I hope the pointer versions fall away.

I could try again though.
I could also try layering handle over pointer in the short term,
but that, arguably, puts off more work.

jaykrell commented:
@monojenkins build Linux i386

jaykrell commented:
@monojenkins Linux x64 Coop GC

jaykrell commented:
@monojenkins Linux x64 mcs

jaykrell commented:
@monojenkins Linux x64 Coop GC

jaykrell commented:
@monojenkins build Linux x64 mcs

jaykrell commented:
@monojenkins build Linux x64 Coop GC

jaykrell commented:
@monojenkins build  Linux x64 mcs

jaykrell commented:
@monolite build deb with monolite

jaykrell commented:
@monojenkins build deb with monolite

lambdageek commented:
@monojenkins build failed

lambdageek commented:
The windows builds are broken: `open_handle` isn't being passed the string length argument.

jaykrell commented:
@monojenkins build Linux x64 mcs

jaykrell commented:
@monojenkins build failed

jaykrell commented:
@monojenkins build failed
